pilot/pkg/model/context.go

```diff
@@ -248,6 +253,43 @@ func GetProxyConfigNamespace(proxy *Proxy) string {
 	return parts[1]
 }
 
+// SetNamespaceDependencies sets the namespaces of the reachable services.
+func (node *Proxy) SetNamespaceDependencies(dependency *meshconfig.MeshConfig_DefaultServiceDependency) {
+	if node == nil || dependency == nil {
+		return
+	}
+	node.NamespaceDependencies = make(map[string]bool)
+	if dependency.ImportMode == meshconfig.MeshConfig_DefaultServiceDependency_ALL_NAMESPACES {
```

Comment:
You need the namespace dependency here as well. The idea is that by default the mesh will have ALL_NAMESPACES as the dependency mode. I can add a network scope to a namespace in order to prune the graph for that namespace. When doing so, you still have to respect the "ImportedNamespaces" in addition to the namespaces declared in the network scope for that namespace.

So you would need to maintain the dependency mode (All/same) per namespace. IF a network scope comes in for a namespace, set a field hasNetworkScope and update the namespace dependencies for this namespace. This namespaceDependencyMap[namespace->[namespaces]] can be maintained in the PushContext and updated whenever serviceDependencies come and go. When doing push.Services(proxy), you can check if the proxy hasServiceDependency, and if so, return the services in the namespaces in namespaceDependencyMap[proxy.ConfigNamespace].

Optimization: have a map of namespace->{services, virtualServices, DestinationRules} in pushContext. So that, when responding to push.Services(proxy), you can get list of namespaceDependencies, and concatenate the services for each namespace.

Reply:
Without ConfigScope this must be good, but if `ConfigScope` set to `Private`. Then another map namespace--->ns private  {services, virtualServices, DestinationRules} may be needed.

Reply:
Yes IC.  Would update in a following pr.